pkg/sfu/buffer/dependencydescriptorparser.go

```diff
@@ -12,91 +13,143 @@ import (
-		}
-		if ddVal.AttachedStructure != nil && !ddVal.FirstPacketInFrame {
-			// r.logger.Debugw("ignoring non-first packet in frame with attached structure")
-			return nil, videoLayer, nil
+	var ddVal dd.DependencyDescriptor
+	ext := &dd.DependencyDescriptorExtension{
+		Descriptor: &ddVal,
+		Structure:  r.structure,
+	}
+	_, err := ext.Unmarshal(ddBuf)
+	if err != nil {
+		// r.logger.Debugw("failed to parse generic dependency descriptor", "err", err, "payload", pkt.PayloadType, "ddbufLen", len(ddBuf))
+		return nil, videoLayer, err
+	}
+
+	if ddVal.FrameDependencies != nil {
+		videoLayer.Spatial, videoLayer.Temporal = int32(ddVal.FrameDependencies.SpatialId), int32(ddVal.FrameDependencies.TemporalId)
+	}
+	if ddVal.AttachedStructure != nil && !ddVal.FirstPacketInFrame {
+		// r.logger.Debugw("ignoring non-first packet in frame with attached structure")
+		return nil, videoLayer, nil
+	}
+
+	if ddVal.AttachedStructure != nil {
+		r.structure = ddVal.AttachedStructure
+		r.decodeTargets = ProcessFrameDependencyStructure(ddVal.AttachedStructure)
+		if len(r.decodeTargets) != 0 {
+			r.logger.Debugw(fmt.Sprintf("update decode targets: %v", r.decodeTargets))
+			go r.onMaxLayerChanged(r.decodeTargets[0].Layer.Spatial, r.decodeTargets[0].Layer.Temporal)
 		}
+	}
 
-		if ddVal.AttachedStructure != nil {
-			var maxSpatial, maxTemporal int32
-			r.structure = ddVal.AttachedStructure
-			r.decodeTargetLayer = r.decodeTargetLayer[:0]
-			for target := 0; target < r.structure.NumDecodeTargets; target++ {
-				layer := VideoLayer{0, 0}
-				for _, t := range r.structure.Templates {
-					if t.DecodeTargetIndications[target] != dd.DecodeTargetNotPresent {
-						if layer.Spatial < int32(t.SpatialId) {
-							layer.Spatial = int32(t.SpatialId)
-						}
-						if layer.Temporal < int32(t.TemporalId) {
-							layer.Temporal = int32(t.TemporalId)
-						}
-					}
-				}
-				if layer.Spatial > maxSpatial {
-					maxSpatial = layer.Spatial
+	if ddVal.AttachedStructure != nil && ddVal.FirstPacketInFrame {
+		r.logger.Debugw(fmt.Sprintf("parsed dependency descriptor\n%s", ddVal.String()))
+	}
+
+	if mask := ddVal.ActiveDecodeTargetsBitmask; mask != nil {
+		var maxSpatial, maxTemporal int32
+		for _, dt := range r.decodeTargets {
+			if *mask&(1<<dt.Target) != uint32(dd.DecodeTargetNotPresent) {
+				if maxSpatial < dt.Layer.Spatial {
+					maxSpatial = dt.Layer.Spatial
 				}
-				if layer.Temporal > maxTemporal {
-					maxTemporal = layer.Temporal
+				if maxTemporal < dt.Layer.Temporal {
+					maxTemporal = dt.Layer.Temporal
 				}
-				r.decodeTargetLayer = append(r.decodeTargetLayer, layer)
 			}
-			r.logger.Debugw("max layer changed", "maxSpatial", maxSpatial, "maxTemporal", maxTemporal)
-			go r.onMaxLayerChanged(maxSpatial, maxTemporal)
 		}
+		r.logger.Debugw("max layer changed", "maxSpatial", maxSpatial, "maxTemporal", maxTemporal)
+		// DD-TODO: should this be called in a goroutine as the preivous invocation above is?
```

Comment:
Added this comment as there is a callback ☝️ also and that uses a go routine, but this does not. I am guessing we will need synchronous so that fps calc works properly, i. e. it does not prematurely declare that it is done? Not sure, leaving a comment for now.

Reply:
I think I know what. We can use DD as a stream tracker (we should change to use DD as stream tracker when using DD as that is client telling what it is going to stop/start). If we use it, that callback will go all the way to all the down tracks and stream allocator etc. But, we are not using it now.  So, will remove the go routine now and we can do a go routine in the callback when we move to DD based stream tracker.